tools/role_permission_configurator.php

```diff
@@ -0,0 +1,516 @@
+        $newRoles = $roleObject->getRolesFromPermissions($userPermissionIDs);
+
+        if (!empty($newRoles)) {
+            echo "Updating roles for $userName\n";
+            // update their roles in the database
+            foreach ($newRoles as $roleID => $roleName) {
+                echo "\tAdding $roleName\n";
+                $DB->insert(
+                    'user_role_rel',
+                    array(
+                        'UserID' => $userID,
+                        'RoleID' => $roleID,
+                    )
+                );
+            }
+            echo "\n";
+        }
+    }
+}
+
+/**
+ * Rebuilds permissions based on user roles. Tops up permissions associated with
+ * each user by looking at which roles the user is associated with. This function
+ * does NOT delete any permissions as some permissions are given
+ * independently from roles
+ *
+ * @return void
+ */
+function rebuildPermissions()
+{
+    global $permissionObject;
+    $DB = \Database::singleton();
+
+    echo "Updating permissions based on user roles\n";
+    echo "Note: This command does not remove any existing user permissions,\t
+        it simply adds to them.\n\n";
+    // Iterate over each user, update their roles based on their permissions
+    foreach (getUsers() as $userID=>$userName) {
+        //Avoid deleting current permissions since
+        //some permissions might not be associated with roles
+
+        // get user roles
+        $roleIDs = getUserRoleIDs($userID);
+
+        // recalculate their roles
+        $newPermissions = $permissionObject->getPermissionsFromRoles($roleIDs);
+
+        //unset existing permissions to avoid insert query
+        foreach ($newPermissions as $permissionID=>$permissionName) {
+            if (userHasPermission($userID, $permissionID)) {
+                unset($newPermissions[$permissionID]);
+            }
+        }
+
+        if (!empty($newPermissions)) {
+            echo "Updating permissions for $userName\n";
+            foreach ($newPermissions as $permissionID=>$permissionName) {
+                if (!userHasPermission($userID, $permissionID)) {
+                    echo "\tAdding $permissionName\n";
+                    $DB->insert(
+                        'user_perm_rel',
+                        array(
+                            'userID' => $userID,
+                            'permID' => $permissionID,
+                        )
+                    );
+                }
+            }
+            echo "\n";
+        }
+    }
+}
+
+/**
+ * Adds a permission to a role. Updates users with that role to either obtain
+ * the new permission or loose the role depending on the $update parameter
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ * @param bool   $update     if true, adds permission to users with current role.
+ *                           if false, removes the role from the user
+ *
+ * @return void
+ */
```

Comment:
I don't see why a "addPermission" function would remove a role from a user.

This portion should be in a "removeRole" function

Reply:
Ok, now I understand, maybe I am slow, or maybe the text was not clear.